### ISSUE TITLE
Train with more data with last.pt or best.pt

### ISSUE THREAD

**Issue by asifmustafa87** [ultralytics/ultralytics#3764]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

Hi,

Thanks for the amazing work. I have a confusion. I trained yolov8(x) on a custom dataset for 300 epochs and it ran successfully for all the epochs.  Now I have more training data and I want to train my model with the new data. Which model should I use in this case? The last.pt or best.pt. Since early stopping did not happen the last.pt should be trained right? Can you please clarify this?

Thanks in advance!

**Comment by glenn-jocher** [comment]
@asifmustafa87 hi there,

Thank you for your kind words and for using YOLOv8! I'm glad to hear that you found it helpful.

To answer your question, when you trained YOLOv8 on your custom dataset for 300 epochs and it completed successfully for all epochs, the `last.pt` checkpoint should contain the weights of the model after the last epoch. This checkpoint represents the state of the model at the end of training.

If you now have more training data and want to continue training your model, you can use the `last.pt` checkpoint as the starting point. The model will continue to learn and update its weights based on the new data. Please note that the training process will start from where it left off, so it will not retrain on the initial dataset again.

On the other hand, the `best.pt` checkpoint is typically used when early stopping occurs during training. It represents the model with the best performance on the validation set. Since you mentioned that early stopping did not happen, the `best.pt` checkpoint may not be relevant in your case.

I hope this clarifies your confusion. If you have any further questions, please feel free to ask. Happy training!

Best regards,
Glenn Jocher

**Comment by asifmustafa87** [comment]
Hi @glenn-jocher ,

Thanks for your detailed answer with all the clarifications. I got the point. 1 last thing. So I can directly put my training data in my training folder and it will train on the new data only? Or should I delete the old train data to start training on the new data? Thanks in advance!

**Comment by bit-scientist** [comment]
@asifmustafa87 If possible, you keep the old data as it helps the model to generalize to both new and old data in case the two happen to be a little different. However, if you don't have the old dataset (which is unlikely in your case) you may just use the new one as long as you don't introduce new classes.

**Comment by glenn-jocher** [comment]
@bit-scientist,

You're welcome! Regarding your question, it really depends on your specific use case. Generally speaking, the more data you can provide to the model, the better it can learn to generalize from that data. Therefore, if you include the original dataset along with the new one, your model can learn from a larger data set and may perform better.

However, if your objective is to train the model strictly on the new dataset, you can choose to train only on that. Just ensure that there's no drift in the class distribution or significant changes from the original data to avoid performance degradation. If you intend to add new classes in your model, you would need to start training from scratch.

In YOLOv8, you can specify the data you want to train on in the configuration file. Simply point to the folder that contains the relevant data. If you want to include both the old and new data, make sure both are in the same folder or directories.

I hope this helps! Let me know if you have any further questions.

Best,
Glenn Jocher.

**Comment by asifmustafa87** [comment]
Thank you @bit-scientist and @glenn-jocher for your help and right direction. Now I have a clear understanding. Have a nice day!

**Comment by glenn-jocher** [comment]
@asifmustafa87, you're welcome! I'm glad we could assist and provide you with a better understanding. If you have any more questions in the future, don't hesitate to reach out. We're always here to help. Have a fantastic day and good luck with your work on YOLOv8!

**Comment by spaceman-cb** [comment]
Hi,

I'd like to clarify the conditions required to train on new data (as opposed to fine-tuning or transfer learning, where where final layers are thrown out and new final layers are trained)

Based on the above, as I understand it:

* if the names and ordering of the classes in the dataset.yaml file do not change, the model will continue to train the existing network instead of fine-tune.
* otherwise, it will fine-tune/do transfer learning/what-have-you

What happens in the following cases, provided the classes stay the same?
* the paths in the dataset.yaml are changed
* a different dataset.yaml file is used (but class names and ordering is preserved)

Thank you!

**Comment by glenn-jocher** [comment]
@spaceman-cb hi,

Great questions!

In YOLOv8, the model training process does not depend specifically on the dataset.yaml file, but rather on the actual labels in your data. The names and order of classes you set in the dataset.yaml file are there mainly for you to manage your data effectively.

When you include new data, the model will continue to train from where it left off with the new data. It will not start over again unless you explicitly reset it. This is true regardless of the specific dataset.yaml file you use, as long as the names and order of classes match with the original training.

If you change the paths in the dataset.yaml file but keep the same classes, the model should handle this without issues, because it uses the paths to find where your data lives and does not use this information to structure the model itself.

In summary, as long as the class structure doesn't change, YOLOv8 will continue training on new data, irrespective of what dataset.yaml file or paths you use.

I hope this explanation helps! Let me know if you have further questions.

Best regards!

**Comment by spaceman-cb** [comment]
@glenn-jocher 
Thank you! That's a huge help and clears up a lot!

**Comment by glenn-jocher** [comment]
@spaceman-cb, 

You're welcome! I'm glad that the information was helpful and cleared up your questions. If you have any more queries regarding YOLOv8, feel free to ask. We're here to help. Happy training! 

Best Regards,
Glenn Jocher

**Comment by Vinaygoudasp7** [comment]
Hi 
i am new to Ai/Ml i am working on yolov8, i have to train this model from scratch so i have selected 3 dataset of different classes. my doubt if i train model from scratch by using yolov8n.yaml by using one of dataset then i train model which i got best.pt (from previous train) with next dataset  can i do like this or if any alternate ways to solve this problem

**Comment by glenn-jocher** [comment]
@Vinaygoudasp7 hi there!

Welcome to the world of AI/ML and YOLOv8! 🚀 It's great to see your enthusiasm for training models from scratch.

For your approach, training sequentially on different datasets can work, but it's generally better to combine all your datasets into one larger dataset if the classes are different. This way, the model learns to generalize across all classes simultaneously, which usually results in better performance.

Here’s a quick example of how you might set up your dataset.yaml to include multiple datasets:

```yaml
# dataset.yaml
train: [path/to/dataset1/images, path/to/dataset2/images, path/to/dataset3/images]
val: [path/to/dataset1/val/images, path/to/dataset2/val/images, path/to/dataset3/val/images]

nc: 10  # number of classes (total from all datasets)
names: ['class1', 'class2', ..., 'class10']
```

Make sure to adjust the paths and class names to fit your specific datasets. Then, you can train your model using this combined dataset configuration.

If you have any more questions or need further assistance, feel free to ask. Happy training! 😊

**Comment by Subhankhan41** [comment]
Hi @glenn-jocher I hope you are doing well. I have a question regarding retraining the model with new datasets or images. Initially, I have trained my custom dataset using the following:
from ultralytics import YOLO
# Load a model
model = YOLO("yolov8n.yaml")  # build a new model from YAML
model = YOLO("yolov8n.pt")  # load a pretrained model (recommended for training)
model = YOLO("yolov8n.yaml").load("yolov8n.pt")  # build from YAML and transfer weights

# Use the model for training
results = model.train(data='../datasets/Rubbish/data.yaml', epochs=100, batch = 16)  # train the model

The best.pt works fine when I deploy the trained model. However, now I have additional 1000 images to train. So, how do I perform that task? I have a new data.yaml for the new dataset for 1000 images, but I am confused that how would I perform the training with my previous best.pt. 

Thanks in advance.

**Comment by glenn-jocher** [comment]
Hi @Subhankhan41,

Thank you for reaching out! I'm glad to hear that your initial training with YOLOv8 went well. To continue training your model with additional data, you can indeed use your `best.pt` from the previous training as a starting point. This approach is often referred to as fine-tuning.

Here's how you can proceed:

1. **Combine Your Datasets**: Ideally, you should combine your original dataset and the new dataset into a single dataset. This ensures that the model learns from all available data simultaneously. Update your `data.yaml` file to include paths to both datasets.

    ```yaml
    # combined_data.yaml
    train: [path/to/original_dataset/images, path/to/new_dataset/images]
    val: path/to/validation_dataset/images  # Optional: You can use a separate validation set

    nc: <number_of_classes>
    names: ['class1', 'class2', ..., 'classN']
    ```

2. **Continue Training**: Load your previous best model and continue training with the combined dataset.

    ```python
    from ultralytics import YOLO

    # Load the previous best model
    model = YOLO("path/to/best.pt")

    # Continue training with the combined dataset
    results = model.train(data='path/to/combined_data.yaml', epochs=50, batch=16)
    ```

If you prefer to train only on the new dataset, you can still use the `best.pt` model as a starting point:

```python
from ultralytics import YOLO

# Load the previous best model
model = YOLO("path/to/best.pt")

# Train with the new dataset
results = model.train(data='path/to/new_data.yaml', epochs=50, batch=16)
```

This approach will fine-tune your model on the new data while retaining the knowledge from the previous training.

If you encounter any issues or have further questions, please feel free to ask. Happy training! 😊

**Comment by codexkunal** [comment]
Hi @glenn-jocher I have trained a yolo8m.yaml model on custom dataset (aerial images). And I have made a web app using Streamlit and now I want to make prediction on new images or on unseen data, then what should I have to use best.pt or last.pt.

**Comment by Subhankhan41** [comment]
Thanks for the response @glenn-jocher, it seems pretty impressive. I tried approaching the following link https://www.ultralytics.com/license to discuss about the Hub enterprise license a week ago, but haven't got any response yet. I would be much obliged if you guys could provide me a response about it. 
Thanks

**Comment by glenn-jocher** [comment]
Thank you for reaching out. For inquiries regarding the Ultralytics Hub enterprise license, please ensure you have used the contact form at [https://www.ultralytics.com/license](https://www.ultralytics.com/license). If you haven't received a response, please try reaching out again through the same link. We handle all licensing inquiries through that channel to ensure proper assistance. Thank you for your understanding.

**Comment by seungbo** [comment]
@glenn-jocher 
<img width="927" alt="스크린샷 2024-09-23 오전 11 30 55" src="https://github.com/user-attachments/assets/c0d97941-e98d-4a71-9b91-84e136d7eda2">
In a model trained with custom data, I added epochs from last.pt model to train and the cls_loss value at 1epoch is higher than the last cls_loss value before training, can you tell me why this is happening?

**Comment by glenn-jocher** [comment]
@seungbo when you resume training, initial fluctuations in loss values, like `cls_loss`, can occur due to learning rate adjustments or changes in data distribution. It's often temporary and should stabilize as training progresses.

**Comment by seungbo** [comment]
@glenn-jocher When I changed workers from 0 to 4, the cls_loss value went down slower than before, any idea why? And I'm currently training the model on windows, so I'd like to run it on otutu, as the environment can affect the training. I would like to try Ubuntu as I understand that the environment can affect the training. Is there any documentation that I can get that I can turn to Ubuntu?

**Comment by glenn-jocher** [comment]
It’s normal for `cls_loss` to fluctuate, especially at the start of training. This can happen due to the model adjusting to new data or changes in learning rate. If it persists, consider reviewing your data and training settings.

**Comment by glenn-jocher** [comment]
When resuming training, an initial increase in `cls_loss` can occur due to adjustments in learning rate or model weights adapting to new data. It's usually temporary and should stabilize as training progresses.

**Comment by glenn-jocher** [comment]
When you resume training, initial fluctuations in loss values can occur due to changes in learning rate or data distribution. It's often temporary and should stabilize as training progresses.

**Comment by seungbo** [comment]
@glenn-jocher How do I use 'resume' when training a model using `yolo segment train data=data.yaml model=yolo11n-seg.yaml epochs=300 imgsz=640 verbose=True lr0=0.001 batch=32 cos_lr=True degrees=5.0 shear=2.0 mixup=0.2 patience=25 plots=True`?
Is this all I need to do: `yolo segment train model=last.pt resume=True`?

**Comment by glenn-jocher** [comment]
Training with additional epochs can initially increase loss due to adjustments in the model's weights. It's common for the loss to fluctuate before stabilizing and decreasing. Continue training to see if the loss improves over time.

**Comment by glenn-jocher** [comment]
It's normal for `cls_loss` to fluctuate initially when resuming training, especially with new data or settings. It should stabilize as training progresses.

**Comment by glenn-jocher** [comment]
When you resume training, initial fluctuations in loss values can occur due to adjustments in learning rate or data differences. It's often temporary and should stabilize as training progresses.

**Comment by glenn-jocher** [comment]
@seungbo it's common for the classification loss to fluctuate initially when resuming training, especially if the learning rate or data distribution has changed. It should stabilize as training progresses.
